### ISSUE TITLE
[BUG]: Error computing HOTA metrics in Carla MOT dpatch defended

### ISSUE THREAD

**Issue by swsuggs** [twosixlabs/armory#1769]
### Description of the bug

I get an error during the final metric calculations when running carla_mot_dpatch_defended.  I'm running on the rc0.16.1 branch.
```IndexError: arrays used as indices must be of integer (or boolean) type```

### Steps To Reproduce

armory run scenario_configs/eval6/carla_mot/carla_mot_dpatch_defended.json

### Additional Information

```
Traceback (most recent call last):
File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
return _run_code(code, main_globals, None,
│         │     └ {'__name__': '__main__', '__doc__': '\nMain script for running scenarios. Users will run a scenario by calling:\n    armory r...
│         └ <code object <module> at 0x7f5a91c72c90, file "/workspace/armory/scenarios/main.py", line 1>
└ <function _run_code at 0x7f5bd99b1d30>
File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
exec(code, run_globals)
│     └ {'__name__': '__main__', '__doc__': '\nMain script for running scenarios. Users will run a scenario by calling:\n    armory r...
└ <code object <module> at 0x7f5a91c72c90, file "/workspace/armory/scenarios/main.py", line 1>
File "/workspace/armory/scenarios/main.py", line 228, in <module>
run_config(
└ <function run_config at 0x7f5a90bfcee0>
File "/workspace/armory/scenarios/main.py", line 144, in run_config
scenario.evaluate()
│        └ <function Scenario.evaluate at 0x7f5a90849430>
└ <armory.scenarios.carla_mot.CarlaMOT object at 0x7f5a90bf3940>
> File "/workspace/armory/scenarios/scenario.py", line 402, in evaluate
self._evaluate()
│    └ <function Scenario._evaluate at 0x7f5a908493a0>
└ <armory.scenarios.carla_mot.CarlaMOT object at 0x7f5a90bf3940>
File "/workspace/armory/scenarios/scenario.py", line 393, in _evaluate
self.finalize_results()
│    └ <function Scenario.finalize_results at 0x7f5a90849310>
└ <armory.scenarios.carla_mot.CarlaMOT object at 0x7f5a90bf3940>
File "/workspace/armory/scenarios/scenario.py", line 381, in finalize_results
self.metric_results = self.metrics_logger.results()
│                     │    │              └ <function MetricsLogger.results at 0x7f5a90a74160>
│                     │    └ <armory.instrument.config.MetricsLogger object at 0x7f5a714c2820>
│                     └ <armory.scenarios.carla_mot.CarlaMOT object at 0x7f5a90bf3940>
└ <armory.scenarios.carla_mot.CarlaMOT object at 0x7f5a90bf3940>
File "/workspace/armory/instrument/config.py", line 118, in results
get_hub().close()
└ <function get_hub at 0x7f5a90ae1310>
File "/workspace/armory/instrument/instrument.py", line 444, in close
meter.finalize()
│     └ <function Meter.finalize at 0x7f5a90af5280>
└ <armory.instrument.instrument.GlobalMeter object at 0x7f5a71642eb0>
File "/workspace/armory/instrument/instrument.py", line 599, in finalize
result = self.final(self._results, **self.final_kwargs)
         │    │
│    │           │    └ {}
         │    │     │    │           └ <armory.instrument.instrument.GlobalMeter object at 0x7f5a71642eb0>
         │    │     │    └ [([{'image_id': 0, 'category_id': 1, 'bbox': [195.0, 253.0, 28.0, 33.0], 'score': 1.0, 'object_id': 1001, 'visibility': 1.0},...
         │    │     └ <armory.instrument.instrument.GlobalMeter object at 0x7f5a71642eb0>
         │    └ <function GlobalMeter.__init__.<locals>.<lambda> at 0x7f5a7164c5e0>
         └ <armory.instrument.instrument.GlobalMeter object at 0x7f5a71642eb0>
File "/workspace/armory/instrument/instrument.py", line 661, in <lambda>
final=lambda x: final_metric(*identity_zip(x), **final_kwargs),
             │  │             │            │     └ {}
             │  │             │            └ [([{'image_id': 0, 'category_id': 1, 'bbox': [195.0, 253.0, 28.0, 33.0], 'score': 1.0, 'object_id': 1001, 'visibility': 1.0},...
             │  │             └ <function identity_zip at 0x7f5a90a72040>
             │  └ <armory.metrics.task.GlobalHOTA object at 0x7f5a71642df0>
             └ [([{'image_id': 0, 'category_id': 1, 'bbox': [195.0, 253.0, 28.0, 33.0], 'score': 1.0, 'object_id': 1001, 'visibility': 1.0},...
File "/workspace/armory/metrics/task.py", line 1733, in __call__
self.hota_metrics.calculate_hota_metrics_per_class_per_video(
│    │            └ <function HOTA_metrics.calculate_hota_metrics_per_class_per_video at 0x7f5a90a739d0>
│    └ <armory.metrics.task.HOTA_metrics object at 0x7f5a71642a00>
└ <armory.metrics.task.GlobalHOTA object at 0x7f5a71642df0>
File "/workspace/armory/metrics/task.py", line 1694, in calculate_hota_metrics_per_class_per_video
] = self.HOTA_calc.eval_sequence(data)
    │    │         │             └ {'gt_ids': [array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]), array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]),...
    │    │         └ <function HOTA.eval_sequence at 0x7f5a70ad41f0>
    │    └ <TrackEval.trackeval.metrics.hota.HOTA object at 0x7f5a71642d00>
    └ <armory.metrics.task.HOTA_metrics object at 0x7f5a71642a00>
File "/armory/tmp/2022-11-21T155526.035195/external/TrackEval/trackeval/_timing.py", line 64, in wrap
return f(*args, **kw)
       │  │       └ {}
       │  └ (<TrackEval.trackeval.metrics.hota.HOTA object at 0x7f5a71642d00>, {'gt_ids': [array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,...
       └ <function HOTA.eval_sequence at 0x7f5a70ad4160>
File "/armory/tmp/2022-11-21T155526.035195/external/TrackEval/trackeval/metrics/hota.py", line 61, in eval_sequence
potential_matches_count[gt_ids_t[:, np.newaxis], tracker_ids_t[np.newaxis, :]
] += sim_iou
│                       │           │  │         │             │  │               └ array([], shape=(12, 0), dtype=float32)
│                       │           │  │         │             │  └ None
│                       │           │  │         │             └ <module 'numpy' from '/opt/conda/lib/python3.8/site-packages/numpy/__init__.py'>
│                       │           │  │         └ array([], dtype=float64)
│                       │           │  └ None
│                       │           └ <module 'numpy' from '/opt/conda/lib/python3.8/site-packages/numpy/__init__.py'>
│                       └ array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])
└ array([[ 0.,  1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
           0.,  0.,  0.,  0.,  0.],
         [ 0.,  0., ...
IndexError: arrays used as indices must be of integer (or boolean) type
```

**Comment by davidslater** [comment]
Is this just on the defended MOT? Or on all of them?

**Comment by swsuggs** [comment]
I've seen it only on the one config specified.  Possibly it happens at random and I got lucky on the others, although it has happened every time I've run this config.

**Comment by davidslater** [comment]
My guess is that this is an edge case where the model doesn't predict anything in a given frame, (`tracker_ids_t is []`), and so runs into problems.

**Comment by davidslater** [comment]
That would be consistent with a defended model being less confident on making predictions.

**Comment by davidslater** [comment]
As an example:
```
>>> import numpy as np
>>> x = np.random.random(10)
>>> x
array([0.27423765, 0.04868633, 0.31757498, 0.42565926, 0.25206396,
       0.63660789, 0.62044409, 0.37587597, 0.11691465, 0.21264816])
>>> index = np.array([])
>>> index
array([], dtype=float64)
>>> x[index]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: arrays used as indices must be of integer (or boolean) type
>>> index = np.array([], dtype=int)
>>> x[index]
array([], dtype=float64)
```

**Comment by swsuggs** [comment]
That would make sense, a similar thing happened with poison filtering a while back

**Comment by swsuggs** [comment]
The ```tracker_ids``` are already initialized ```astype(int)``` ([link](https://github.com/twosixlabs/armory/blob/2efa5912f351564d2879053653a9559c50e97ada/armory/metrics/task.py#L1648)),
but apparently some other subsequent operations change the type at least of any empty arrays.

One option is just to cycle through ```data["tracker_ids"]``` right before the return, and cast them all back to int types:
```
# armory/metrics/task.py
1682      . . . 
1683      for t in range(len(data["tracker_ids"])):
1684          data["tracker_ids"][t] = data["tracker_ids"][t].astype(int)
1685
1686      return data

```
This seemed to solve the problem in my initial test.  @davidslater any drawbacks you see with this approach?

**Comment by jprokos26** [comment]
I believe the error is coming from [this line](https://github.com/twosixlabs/armory/blob/rc0.16.1/armory/metrics/task.py#L1595) since an empty array defaults to float64. Working on a fix now and will make a PR.

**Comment by swsuggs** [comment]
Fixed by #1772
